## Reading CSV file


In [4]:
import pandas as pd
df= pd.read_csv('E:\\Internship\\Book1.csv')
df


Company  Fiscal Year  Total Revenue  Net Income  Total Assets  \
0         NaN          NaN            NaN         NaN           NaN   
1   Microsoft       2021.0       168088.0     61271.0      333779.0   
2   Microsoft       2022.0       198270.0     72738.0      364840.0   
3   Microsoft       2023.0       211915.0     72361.0      411976.0   
4         NaN          NaN            NaN         NaN           NaN   
5       Tesla       2021.0        53823.0      5644.0       62131.0   
6       Tesla       2022.0        81462.0     12587.0       82338.0   
7       Tesla       2023.0        96773.0     14974.0      106618.0   
8         NaN          NaN            NaN         NaN           NaN   
9       Apple       2021.0       365817.0     94680.0      351002.0   
10      Apple       2022.0       394328.0     99803.0      352755.0   
11      Apple       2023.0       383285.0     96995.0      352583.0   

    Total Liabilities  Cash Flow from Operating Activities  
0                 NaN                                  NaN  
1            191791.0                              76740.0  
2            198298.0                              89035.0  
3            205753.0                              87582.0  
4                 NaN                                  NaN  
5             30548.0                              11497.0  
6             36440.0                              14724.0  
7             43009.0                              13256.0  
8                 NaN                                  NaN  
9            287912.0                              43887.0  
10           302083.0                              51345.0  
11           290437.0                              54847.0

## Calculation of year-over-year changes

In [17]:
df['Revenue Growth (%)'] = df.groupby(['Company'])['Total Revenue'].pct_change() * 100
df['Net Income Growth (%)'] = df.groupby(['Company'])['Net Income'].pct_change() * 100
df['Assets Growth (%)'] = df.groupby(['Company'])['Total Assets'].pct_change() * 100
df['Liabilities Reduction (%)'] = df.groupby(['Company'])['Total Liabilities'].pct_change() * 100
df['Cash Flow Growth (%)'] = df.groupby(['Company'])['Cash Flow from Operating Activities'].pct_change() * 100

df.dropna(how='all')

Company  Fiscal Year  Total Revenue  Net Income  Total Assets  \
1   Microsoft       2021.0       168088.0     61271.0      333779.0   
2   Microsoft       2022.0       198270.0     72738.0      364840.0   
3   Microsoft       2023.0       211915.0     72361.0      411976.0   
5       Tesla       2021.0        53823.0      5644.0       62131.0   
6       Tesla       2022.0        81462.0     12587.0       82338.0   
7       Tesla       2023.0        96773.0     14974.0      106618.0   
9       Apple       2021.0       365817.0     94680.0      351002.0   
10      Apple       2022.0       394328.0     99803.0      352755.0   
11      Apple       2023.0       383285.0     96995.0      352583.0   

    Total Liabilities  Cash Flow from Operating Activities  \
1            191791.0                              76740.0   
2            198298.0                              89035.0   
3            205753.0                              87582.0   
5             30548.0                              11497.0   
6             36440.0                              14724.0   
7             43009.0                              13256.0   
9            287912.0                              43887.0   
10           302083.0                              51345.0   
11           290437.0                              54847.0   

    Revenue Growth (%)  Net Income Growth (%)  Assets Growth (%)  \
1                  NaN                    NaN                NaN   
2            17.956071              18.715216           9.305858   
3             6.882030              -0.518299          12.919636   
5                  NaN                    NaN                NaN   
6            51.351653             123.015592          32.523217   
7            18.795267              18.964010          29.488207   
9                  NaN                    NaN                NaN   
10            7.793788               5.410858           0.499427   
11           -2.800461              -2.813543          -0.048759   

    Liabilities Reduction (%)  Cash Flow Growth (%)  
1                         NaN                   NaN  
2                    3.392756             16.021631  
3                    3.759493             -1.631942  
5                         NaN                   NaN  
6                   19.287678             28.068192  
7                   18.026894             -9.970117  
9                         NaN                   NaN  
10                   4.921990             16.993643  
11                  -3.855232              6.820528

# Conclusion
1. There is a reduction in net income growth and cash flow growth in Microsoft in 2023 by 0.518299% and 1.631942 respectively.
2. Tesla's cash flow growth reduced by 9.970117% in 2023.
3. Apple suffer loss in terms of revenue growth, net income growth and assets growth by 2.800461%, 2.813543% and 0.048759% respectively in 2023.
4. Apple's liabilities got reduced by 3.855232% in 2023 which is good for them.